### DISCUSSION TITLE
[C++] Need help to retain taint in class objects if we don't know the order of execution

### DISCUSSION THREAD

**Discussion by marpom** [github/codeql#16691] [Q&A]
Hello,

This is probably a C++ edge case. Assume that we have an object that exposes two functions and we can't know which function will be called first. In this case, I would like anything that gets tainted in one function to retain its taint when accessed by the other (even if it leads to over-tainting).

For instance:
`library.cc`:
```
#include "absl/base/casts.h"

void outside_fun(uint64_t arg);

class InnerClass {
  public:
   uint64_t tainted = 0;
   uint64_t not_tainted = 0;
};

class ExampleClass {
 private:
  InnerClass obj;
  void HandleWrite(uint64_t value) {
    obj.tainted = value;
    obj.not_tainted = 1;
    uint64_t tmp = obj.tainted;
    outside_fun(tmp);
  }

  void test() {
    uint64_t tmp = obj.tainted;
    outside_fun(tmp);
    tmp = obj.not_tainted;
    outside_fun(tmp);
  }
};
```
`BUILD`:
```
cc_library(
    name = "library",
    srcs = ["library.cc"],
    deps = [
        "@com_google_absl//absl/base:base",
    ],
)
```
and `query.ql`:
```
import cpp
import semmle.code.cpp.dataflow.new.DataFlow
import semmle.code.cpp.dataflow.new.TaintTracking

module VanadiumConfig implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node source) { 
    exists( Function function | 
      function.getName() = "HandleWrite" and
      source.asParameter() = function.getParameter(0)
    )
  }

  predicate isSink(DataFlow::Node sink) {
    any()
  }
}


module VanadiumFlow = TaintTracking::Global<VanadiumConfig>;
import VanadiumFlow::PathGraph


from VanadiumFlow::PathNode source, VanadiumFlow::PathNode sink
where VanadiumFlow::flowPath(source, sink)
select source, source.getLocation(), sink, sink.getLocation()
```

Create the db with:
`codeql database create database --language=cpp --command='bazel build --spawn_strategy=local --nouse_action_cache --noremote_accept_cached --noremote_upload_local_results //library:library'`

and execute the query with:
`codeql query run --database=`\<path\>`/database --  `\<path\>`/query.ql`

In this example, the module identifies the tainted member at line 17, taints the variable and also identifies the flow at line 18. This does not happen at lines 22-23 which is what I'm trying to do (note that if we add a call to `test()` after line 18, the results do include lines 22-23).

Does anyone have any ideas on how to modify the `isAdditionalFlowStep()` predicate to capture this case?

Thanks!

**Comment by MathiasVP** [comment]
Hmm... This is hard for a couple of reasons. I don't have a good solution off the top of my head. Let me try to explain why it's hard:

My first thought was to add an additional flow step from the final (implicit) mention of `this` in `HandleWrite` to the implicit definition of `this` in `test`. That is, we could transfer flow from:
```qll
void HandleWrite(uint64_t value) {
  obj.tainted = value;
  obj.not_tainted = 1;
  uint64_t tmp = obj.tainted; // <-- `this` is implicitly accessed here
  outside_fun(tmp);
}
```
to
```ql
void test() {
  // <-- `this` is implicitly defined here
  uint64_t tmp = obj.tainted;
  outside_fun(tmp);
  tmp = obj.not_tainted;
  outside_fun(tmp);
}
```
However, a non-documented requirement is that the "access path" (i.e., the list of fields that must be accessed to read the data we're tracking) must be empty for any additional flow steps to be taken into account during the analysis.

This is a problem since the access path is `[obj, tainted]` at the point of the last mention of the implicit `this` in `HandleWrite` (because the value being tracked has been written to `this->obj.tainted`). So no matter what we add in `isAdditionalFlowStep` those steps won't actually be taken into account after the first write to `obj.tainted` in `HandleWrite` and until there's been a corresponding read of `obj.tainted` along the path.

Now, if we're following flow into `outside_fun` that's not a problem since there is a read of `obj.tainted` along that path (namely the path that reads `obj.tainted` and stores the result into `tmp` in `HandleWrite`). However, if we want to transfer the flow into `test` that's not the path we want to take. So we need to transfer the flow to `test` without reading the reducing the access path to the empty list (so that we can read if into the `tmp` variable in the `test` member function).

So let's assume that we have a way of adding this flow step from the final (implicit) mention of `this` in `HandleWrite` to the implicit definition of `this` in `test` in a way that makes it usable even without an empty access path. This means it needs to be a _value preserving step_. For C/C++ it's not currently possible to add such steps outside the inner working of the dataflow library ([localFlowStep](https://github.com/github/codeql/blob/32260e2823a0b83c019c78ffda9437d4f22c21dd/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll#L2161) contains all those steps, although that is deep within the library and not expected to be touched by people without wizard-level knowledge).

_However_! Now we run into another invariant: Value preserving steps must preserve the enclosing callable. In order words: Dataflow steps that can be taken while the access path is non-empty cannot transfer flow from one function to another. That's a problem in our case since we're specifically looking for a way to transfer flow from `HandleWrite` to `test`.

So, in summary, I'm not really sure how to add such a step (either via `isAdditionalFlowStep` or by hardcoding it deep inside the dataflow library itself). I'll ask around for other wizard-level ideas, but at the moment I'm not sure there's a good solution.

**Comment by geoffw0** [comment]
> For C/C++ it's not currently possible to add such steps outside the inner working of the dataflow library

It is possible to add value steps using models-as-data (as you can see in the implementation of [`localFlowStep`](https://github.com/github/codeql/blob/32260e2823a0b83c019c78ffda9437d4f22c21dd/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll#L2161)), though (1) I'm not sure if this feature is in the current or next release and (2) it doesn't get you around the second invariant @mathiasvp describes above, that the source and sink with an access path are in the same callable.

---

Can we taint the `ExampleClass` object itself as a proxy for tainting `ExampleClass.obj.tainted`?  So the goal would be for `HandleWrite()` to taint `this` rather than `this->obj.tainted`, then use `isAdditionalFlowStep` to transfer taint interprocedurally to accesses of `ExampleClass.obj.tainted`?
